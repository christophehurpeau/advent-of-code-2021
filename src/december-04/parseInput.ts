export interface Board {
  rows: number[][];
  columns: number[][];
}

export interface BingoGame {
  numbersToDraw: number[];
  boards: Board[];
}

export const parseInput = (input: string): BingoGame => {
  const lines = `${input.trim()}\n`.split('\n');
  const numbersToDraw = lines
    .shift()!
    .split(',')
    .map((v) => parseInt(v, 10));
  const boards: Board[] = [];
  let rows: number[][] = [];

  lines.shift();

  for (const line of lines) {
    if (line === '') {
      boards.push({
        rows,
        // eslint-disable-next-line @typescript-eslint/no-loop-func
        columns: rows[0].map((_, index) => {
          return rows.map((row) => row[index]);
        }),
      });
      rows = [];
    } else {
      rows.push(
        line
          .trim()
          .split(/\s+/)
          .map((v) => parseInt(v, 10)),
      );
    }
  }

  return { numbersToDraw, boards };
};
