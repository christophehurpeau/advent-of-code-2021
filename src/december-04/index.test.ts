import fs from 'fs';
import { parseInput } from './parseInput';
import { findWinningBoard, findLoosingBoard } from '.';

const input = fs.readFileSync('./src/december-04/input.txt', 'utf-8');

const exampleInput = `
7,4,9,5,11,17,23,2,0,14,21,24,10,16,13,6,15,25,12,22,18,20,8,19,3,26,1

22 13 17 11  0
 8  2 23  4 24
21  9 14 16  7
 6 10  3 18  5
 1 12 20 15 19

 3 15  0  2 22
 9 18 13 17  5
19  8  7 25 23
20 11 10 24  4
14 21 16 12  6

14 21 17 24  4
10 16 15  9 19
18  8 23 26 20
22 11 13  6  5
 2  0 12  3  7
`;

describe('part1', () => {
  test('parseInput', () => {
    expect(parseInput(exampleInput)).toMatchSnapshot();
  });

  test('example', () => {
    const res = findWinningBoard(exampleInput);
    expect(res).toBe(4512);
  });

  test('input', () => {
    const res = findWinningBoard(input);
    expect(res).toMatchInlineSnapshot('33462');
  });
});

describe('part2', () => {
  test('example', () => {
    const res = findLoosingBoard(exampleInput);
    expect(res).toBe(1924);
  });

  test('input', () => {
    const res = findLoosingBoard(input);
    expect(res).toMatchInlineSnapshot('30070');
  });
});
