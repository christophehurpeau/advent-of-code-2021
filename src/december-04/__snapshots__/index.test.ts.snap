// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`part1 parseInput 1`] = `
Object {
  "boards": Array [
    Object {
      "columns": Array [
        Array [
          22,
          8,
          21,
          6,
          1,
        ],
        Array [
          13,
          2,
          9,
          10,
          12,
        ],
        Array [
          17,
          23,
          14,
          3,
          20,
        ],
        Array [
          11,
          4,
          16,
          18,
          15,
        ],
        Array [
          0,
          24,
          7,
          5,
          19,
        ],
      ],
      "rows": Array [
        Array [
          22,
          13,
          17,
          11,
          0,
        ],
        Array [
          8,
          2,
          23,
          4,
          24,
        ],
        Array [
          21,
          9,
          14,
          16,
          7,
        ],
        Array [
          6,
          10,
          3,
          18,
          5,
        ],
        Array [
          1,
          12,
          20,
          15,
          19,
        ],
      ],
    },
    Object {
      "columns": Array [
        Array [
          3,
          9,
          19,
          20,
          14,
        ],
        Array [
          15,
          18,
          8,
          11,
          21,
        ],
        Array [
          0,
          13,
          7,
          10,
          16,
        ],
        Array [
          2,
          17,
          25,
          24,
          12,
        ],
        Array [
          22,
          5,
          23,
          4,
          6,
        ],
      ],
      "rows": Array [
        Array [
          3,
          15,
          0,
          2,
          22,
        ],
        Array [
          9,
          18,
          13,
          17,
          5,
        ],
        Array [
          19,
          8,
          7,
          25,
          23,
        ],
        Array [
          20,
          11,
          10,
          24,
          4,
        ],
        Array [
          14,
          21,
          16,
          12,
          6,
        ],
      ],
    },
    Object {
      "columns": Array [
        Array [
          14,
          10,
          18,
          22,
          2,
        ],
        Array [
          21,
          16,
          8,
          11,
          0,
        ],
        Array [
          17,
          15,
          23,
          13,
          12,
        ],
        Array [
          24,
          9,
          26,
          6,
          3,
        ],
        Array [
          4,
          19,
          20,
          5,
          7,
        ],
      ],
      "rows": Array [
        Array [
          14,
          21,
          17,
          24,
          4,
        ],
        Array [
          10,
          16,
          15,
          9,
          19,
        ],
        Array [
          18,
          8,
          23,
          26,
          20,
        ],
        Array [
          22,
          11,
          13,
          6,
          5,
        ],
        Array [
          2,
          0,
          12,
          3,
          7,
        ],
      ],
    },
  ],
  "numbersToDraw": Array [
    7,
    4,
    9,
    5,
    11,
    17,
    23,
    2,
    0,
    14,
    21,
    24,
    10,
    16,
    13,
    6,
    15,
    25,
    12,
    22,
    18,
    20,
    8,
    19,
    3,
    26,
    1,
  ],
}
`;
